File: library/src/main/java/com/shxhzhxx/sdk/ui/LoadMoreAdapter.java
package com.shxhzhxx.sdk.ui;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.shxhzhxx.sdk.R;

public class LoadMoreAdapter<VH extends RecyclerView.ViewHolder> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int LOADING_LAYOUT_ID = R.layout.list_loading;
    private RecyclerView.Adapter<RecyclerView.ViewHolder> mWrapped;
    private boolean mVisible = false;

    public LoadMoreAdapter(RecyclerView.Adapter<RecyclerView.ViewHolder> wrapped) {
        mWrapped = wrapped;
        mWrapped.registerAdapterDataObserver(new DataSetChangeDelegate());
    }

    public void setLoadingVisible(boolean visible) {
        if (mVisible != visible) {
            mVisible = visible;
            if (visible) {
                notifyItemInserted(getItemCount() - 1);
            } else {
                notifyItemRemoved(getItemCount());
            }
        }
    }

    public boolean isLoadingVisible() {
        return mVisible;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == LOADING_LAYOUT_ID)
            return new LoadingHolder(LayoutInflater.from(parent.getContext()).inflate(LOADING_LAYOUT_ID, parent, false));
        return mWrapped.onCreateViewHolder(parent, viewType);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (position < mWrapped.getItemCount()) {
            mWrapped.onBindViewHolder(holder, position);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (mVisible && position == getItemCount() - 1)
            return LOADING_LAYOUT_ID;
        return mWrapped.getItemViewType(position);
    }

    @Override
    public int getItemCount() {
        return mWrapped.getItemCount() + (mVisible ? 1 : 0);
    }

    class LoadingHolder extends RecyclerView.ViewHolder {
        LoadingHolder(View itemView) {
            super(itemView);
        }
    }

    private class DataSetChangeDelegate extends RecyclerView.AdapterDataObserver {

        @Override
        public void onChanged() {
            notifyDataSetChanged();
        }

        @Override
        public void onItemRangeRemoved(int positionStart, int itemCount) {
            notifyItemRangeRemoved(positionStart, itemCount);
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount) {
            notifyItemRangeChanged(positionStart, itemCount);
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
            notifyItemRangeChanged(positionStart, itemCount, payload);
        }

        @Override
        public void onItemRangeInserted(int positionStart, int itemCount) {
            notifyItemRangeInserted(positionStart, itemCount);
        }

        @Override
        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
            notifyItemMoved(fromPosition, toPosition);
        }
    }
}

File: library/src/main/java/com/shxhzhxx/sdk/ui/LoadMoreAdapter.kt
package com.shxhzhxx.sdk.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.shxhzhxx.sdk.R

private val LOADING_LAYOUT_ID = R.layout.list_loading

private class LoadingHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
class LoadMoreAdapter<VH : RecyclerView.ViewHolder>(private val mWrapped: RecyclerView.Adapter<VH>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    var isLoadingVisible = false
        set(visible) {
            if (isLoadingVisible != visible) {
                field = visible
                if (visible) {
                    notifyItemInserted(itemCount - 1)
                } else {
                    notifyItemRemoved(itemCount)
                }
            }
        }

    init {
        mWrapped.registerAdapterDataObserver(DataSetChangeDelegate())
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {

        return if (viewType == LOADING_LAYOUT_ID) LoadingHolder(LayoutInflater.from(parent.context).inflate(LOADING_LAYOUT_ID, parent, false)) else mWrapped.onCreateViewHolder(parent, viewType)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position < mWrapped.itemCount) {
            @Suppress("UNCHECKED_CAST")
            mWrapped.onBindViewHolder(holder as VH, position)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isLoadingVisible && position == itemCount - 1) LOADING_LAYOUT_ID else mWrapped.getItemViewType(position)
    }

    override fun getItemCount(): Int {
        return mWrapped.itemCount + if (isLoadingVisible) 1 else 0
    }

    private inner class DataSetChangeDelegate : RecyclerView.AdapterDataObserver() {

        override fun onChanged() {
            notifyDataSetChanged()
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            notifyItemRangeRemoved(positionStart, itemCount)
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            notifyItemRangeChanged(positionStart, itemCount)
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
            notifyItemRangeChanged(positionStart, itemCount, payload)
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            notifyItemRangeInserted(positionStart, itemCount)
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            notifyItemMoved(fromPosition, toPosition)
        }
    }
}
